### PR TITLE
Deprecate rule set methods which expose implementation details of detekt-core

### PR DESCRIPTION
**Pull request by arturbosch**
Refactorings for #2365 and #2341

### CODE REVIEW

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/RuleSetProvider.kt [hunk 1]
```diff
@@ -22,6 +22,8 @@ interface RuleSetProvider {
      * Api notice: As the rule set id is not known before creating the rule set instance,
      * we must first create the rule set and then check if it is active.
      */
+    @Suppress("DEPRECATION")
+    @Deprecated("Exposes detekt-core implementation details.")
     fun buildRuleset(config: Config): RuleSet? {
```
Comment (BraisGabin) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/RuleSetProvider.kt]: Does it?

Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/RuleSetProvider.kt]: A rule writer does not need to know how his rules will be created and how the active check is done and how we actually use the `includes` and `excludes` to filter if a KtFile should be analyzed. A `RuleSetProvider` just provide the rules.

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/RuleSet.kt [hunk 2]
```diff
@@ -21,12 +21,14 @@ class RuleSet(val id: RuleSetId, val rules: List<BaseRule>) {
...
      */
+    @Deprecated("Exposes detekt-core implementation details.")
     var pathFilters: PathFilters? = null
 
     /**
      * Visits given file with all rules of this rule set, returning a list
      * of all code smell findings.
      */
+    @Deprecated("Exposes detekt-core implementation details.")
     fun accept(file: KtFile, bindingContext: BindingContext = BindingContext.EMPTY): List<Finding> =
```
[the remark this answers is not in the document]
Comment (arturbosch) [reply detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/RuleSet.kt]: The body of the function exposes how detekt internals work.
The user of the api does not need to know how `detekt-core` creates instances of the rules and runs over the KtFile's.

**Comment by arturbosch** [comment]
> I'm trying to understand this refactor. At the end how does we want to `RuleSet` be like?
> 
> just a class like this:
> 
> ```kotlin
> class RuleSet(val id: RuleSetId, val rules: List<BaseRule>, include: String, exclude: String) {
>   private val pathFilters: PathFilters = PathFilters.of(include, exclude)
> 
>   fun shouldAnalyzeFile(file: KtFile): Boolean = //...
> }
> ```
> 
> And nothing else?

If I would rewrite detekt now, I would consider this code for a RuleSetProvider:
```
interface RuleSetProvider {
    val ruleSetId: String

    fun rules(config: Config): List<Rule>
}
```

A rule author does not need to know that we bundle the rules inside a `RuleSet` class which has some logic like `filtering` and `isActive` etc.
He knows that there are some cli flags and config propeties for this but does not need to see how we iterate over rules, visit them and collect the findings :).

**Comment by BraisGabin** [comment]
Oh, OK, so you would remove the rule set from the api all together. That have sense.

**Comment by BraisGabin** [comment]
Oh! We should add information in the depreciation message about what the user should use instead. And, if possible, the regex to do it automatically.

**Comment by arturbosch** [comment]
The problem is the user should not use these methods at all. There is no migration for the user.
He can test his rules one by one. Or if he really wants it he can pull in `detekt-test` and start a `Runner`.

**Pull request merged**